Define CMP v4 API (#134)
guardian/consent-management-platform: README.md

    if (tcfState) {
        console.log(tcfState); // { 1: true || false, 1: true || false, ... }
    }
});
```

## API

### init(options)

returns: `void`

Adds the relevent privacy framework to the page. It needs to be run before any other API call.

#### options.isInUsa

type: `boolean`

_Required!_ Declare whether your user is in the USA or not.

### onConsentNotification(callback)

returns: `void`

#### callback(result)

type: `function`

Callbacks are invoked when the consent state:

-   is acquired
-   changes (eg. if a user changes their privacy preferences).

If the consent state has already been acquired when `onConsentNotification` is called, the callback will be invoked immediately.

##### result.tcfState

type: `Object` or `undefined`

Reports the user's preferences to the TCFv2 purposes:

```js
{
    1: Boolean,
    2: Boolean,
    // etc
}
```

##### result.ccpaState

type: `Object` or `undefined`

Reports whether user has withdrawn consent to sell their data:

```js
{
    doNotSell: Boolean;
}
```

### checkUiWillShow()

returns: `Promise<Boolean>`

The returned promise resolves to `true` if the user will be shown the initial privacy message, or `false` if not.

> If it's called before `init()`, it will `reject`.

Reviewer: does this need to reject? what if it just didn't resolve until `init` has finished? if you never call `init` then `checkUiWillShow` never resolves?

Author: The idea behind rejecting is that it tells the developer using this package that they've forgot to call `init` beforehand. If `init` has been called, even if the initialisation process hasn't finished yet, it resolves. Not resolving it will not make it clear to the developer that something has gone wrong with their implementation.

[![Generic badge](https://img.shields.io/badge/google-chat-259082.svg)](https://chat.google.com/room/AAAAhlhgDTU)

## Usage

```js
import {
    init,
    onConsentNotification,
} from '@guardian/consent-management-platform';

init({ isInUsa: true });

onConsentNotification(({ tcfState, ccpaState }) => {
    if (ccpaState) {
        console.log(ccpaState); // { doNotSell: true || false }
    }
    if (tcfState) {
        console.log(tcfState); // { 1: true || false, 1: true || false, ... }
    }
});
```

## API

### init(options)

returns: `void`

Adds the relevent privacy framework to the page. It needs to be run before any other API call.

#### options.isInUsa

type: `boolean`

_Required! Throws error if not present._ Declare whether your user is in the USA or not.

### onConsentNotification(callback)

returns: `void`

Callbacks are invoked when the consent state:

-   is acquired
-   changes (eg. if a user changes their privacy preferences).

If the consent state has already been acquired when `onConsentNotification` is called, the callback will be invoked immediately.

#### callback(result)

type: `function`

> Reports the user's relevant privacy preferences.

Reviewer: ```suggestion
Reports the user's relevant privacy preferences.

##### result

type: `Object`

```

```js

> {

Reviewer: ```suggestion
##### result

type: `Object` or `undefined`

{
```

    tcfState?: Boolean,
    ccpaState?: Boolean,
}
```

> ###### result.tcfState

Reviewer: ```suggestion
##### result.tcfState
```

type: `Object` or `undefined`

Reports the user's preferences to each of the TCFv2 purposes:

```js
{
    1: Boolean;
    2: Boolean;
    // etc
}
```

> ###### result.ccpaState

Reviewer: ```suggestion
##### result.ccpaState
```

# Consent Management Platform

Welcome to the Consent Management Platform, a library to manage privacy frameworks across \*.theguardian.com, from messaging to reporting the privacy state. All exports include Typescript definitions.

## What utilities does this offer?

### TCF and CCPA privacy frameworks

If you need to use the TCF or CCPA privacy frameworks, you can inititalise them on your page using the `init` function.

### Consent notifications

If you need to conditionally run some code based on a user's consent state you can use the function `onConsentNotification`.

## What is the API?

[init](#init)<br />
[onConsentNotification](#onconsentnotification)<br />
[checkUiWillShow](#checkuiwillshow)<br />
[showPrivacyManager](#showprivacymanager)

### init

`init(options: InitOptions): void`

Calling `init` will add the TCF or the CCPA privacy framework to the page, depending on the configuration options received. We refer to these as TCF mode and CCPA mode, respectively. This function needs to be run before any other API call.

The configuration object that it requires is:

```
interface InitOptions {
	isInUS: boolean;
}
```

> If `InitOptions.isInUS` is missing, `init` will default to running in TCF mode.

Reviewer: this feels dangerous to me – would it be safer to force people to be explicit?

Author: I don't think it's that dangerous because in TCFv2 mode nothing fires unless given explicit consent but I'm happy to change it.
What do you suggest happens if the `isInUS` flag is missing?

Reviewer: I was thinking that because we have to show the CCPA in California, we shouldn't give people the chance to not consider where their code will run?

how about we throw an error, as well as TS failing?

Author: Throwing an error seems good since projects that don't use TS, such as frontend, won't see a TS error.

The signatures for the callback function and its parameters are:

```
interface TcfState {
    [key: string]: boolean
}

type CcpaState = boolean

type ConsentCallback = (tcfSate: TcfState | null, ccpaState: CcpaState | null) => void
```

The keys in `TCFState` will match the TCF purpose IDs.
The value of `ccpaState` will be the reverse of the [CCPA opt-out flag](https://github.com/InteractiveAdvertisingBureau/USPrivacy/blob/master/CCPA/US%20Privacy%20String.md#us-privacy-string-format) when in CCPA mode. This is to provide a homogenous way to handle consent between the TCF and the CCPA frameworks where `true` means consent has been given and `false` means consent has been denied.

If `onConsentNotification` is called before `init`, it will do nothing.

**Example:**

```js
import { onConsentNotification } from '@guardian/consent-management-platform';

onConsentNotification((tcfState, ccpaState) => {
	// Check whether it's in TCF or CCPA mode
	if (tcfState !== null) {
		console.log(tcfState); // { 1: true || false, 1: true || false, ... }
	} else {

> 		console.log(tcfState); // true || false

Reviewer 2: typo here I assume you mean log `ccpaState` in this else block

## Usage

```js
import {
    init,
    onConsentNotification,
} from '@guardian/consent-management-platform';

> init({ isInUsa: true });

Reviewer: :+1:

type: `boolean`

> _Required! Throws error is not present._ Declare whether your user is in the USA or not.

Reviewer: ```suggestion
_Required! Throws error if not present._ Declare whether your user is in the USA or not.
```

type: `Object` or `undefined`

> Reports the user's preferences to the TCFv2 purposes:

Reviewer: cheers :)

```js
{

>     doNotSell: Boolean;

Reviewer: the example above uses js-style `,` – we should probably match them. want to go with TS version?

Author: The linter added this. But I supposed if it is a type definition TS makes more sense.

checkUiWillShow()
    .then(result =>
        console.log(result); // true || false
    ).catch(e =>
        console.log("checkUiWillShow() failed:", e); // "checkUiWillShow() failed: called before init()"
    );
```

### showPrivacyManager()

returns: `void`

> Surfaces the relevant privacy manager, allowing the user to change their privacy preferences after the initial privacy message.

Reviewer: 'surfaces' reads a bit strangely here to me...

```suggestion
Shows the relevant privacy manager, allowing the user to update their privacy preferences after they've closed the initial privacy message.
```
